### DISCUSSION TITLE
09/2023 - Ongoing initiatives

### DISCUSSION THREAD

**Discussion by nikolasrieble** [recharts/recharts#3727] [Announcements]
# Roadmap

This discussion aims at providing at high level overview of ongoing projects. These are the projects where our focus lies, and where we will want to spend most of our time. 

That being said, if you want to contribute to other topics or any open issues, you are welcome to do so.

# Focus points / Projects
Anyone who wants to help us will find greatest support along the lines of these projects. We believe that solving these problems will put recharts (back) on the road to continuous success.

## Documentation - Storybook
The website has an API section and an example section. Both sections are horribly outdated and partially misleading. In some cases props are missing, in other cases types are incomplete. 

The effort to maintain documentation up to date is high. Thus we want to move to storybook instead, where we can have live examples, automatically up to date, since auto-generated from master code. 

Our effort to move to Storybook is in progress. You can find open issues under the label [storybook](https://github.com/recharts/recharts/labels/storybook).

## Operational Excellence - Bundle Size, Maintainability, Readability
Recharts is [quite big](https://github.com/recharts/recharts/issues/3697). One of the road blocks for improving bundle size is the fact that most charts have a single entry point, `generateCategoricalChart`. It is a single large component that has a lot of behaviour implemented which most used is unused (i.e. the BarChart uses generateCategoricalChart, but never renders a RadialAxis). 

This component is also difficult to understand, test and maintain. Thus we want to decouple and simplify parts of it. 

We are discussing our strategy to attack this large file [here](https://github.com/recharts/recharts/discussions/3717). If you have ideas, please let us know. What we are looking for here are small steps, that are non breaking and help us simplify the code base. 

## D3-Scale / Victory-vendor
To solve a security vulnerability, we had to upgrade d3-scale. That turned out to be rather complex. We instead decided to use [victory-vendor](https://github.com/recharts/recharts/pull/3167). You can find more information about the approach and the root cause [here](https://formidable.com/blog/2022/victory-esm/). Using victory-vendor was only a short term solution to an urgent need. 

Now we want to revert this an properly fix it. The upgrade of d3-scale breaks our jest setup, and we are currently investigating whether switching to vitest is solving our problem. 
Here, we are facing yet other tests to fail, but it overall still seems like a reasonable approach.

Once we can get all tests to work with vitest https://github.com/recharts/recharts/pull/3705, we will migrate from jest to vitest, and then upgrade d3-scale. 

# Major release 3.x
Given the projects about the next release will most likely only contain fixes to larger issues that are breaking. We are currently not targeting any large new features. 

We might want to:
- evaluate and improve on the types (i.e. less / no any) 
- unify the API across components
TBD

**Comment by nikolasrieble** [comment]
@ckifer This is supposed to be a single high level entry point to current initiatives, linking to and leading to more detailed pages per initiative. What is this missing so far?

**Reply by ckifer** [reply]
> Documentation - Storybook

Looks good

> Bundle size - Refactoring

This isn't 100% about bundle size - maybe we call it Operational Excellence with mentions of bundle size, maintainability, readability, etc?

Maybe we should add:
1. section on addressing issues that have the most reactions/are longest standing known problems
2. section on 3.x - what that means, what it doesn't mean, why we need a major release
3. potentially mention something about `victory-vendor`?

**Reply by nikolasrieble** [reply]
Great, thank you. I iterated and added a section about victory-vendor, and added a few thoughts to 3.x Please do extend / comment.
